kvserver: simplify and document handling of ReplicaPlaceholder

tbg opened the pull request:
This commit cleans up the handling of ReplicaPlaceholder and documents their
semantics better. I embarked on this while looking into an unrelated test
flake. It occurred to me that we were not installing placeholders before the
snapshot transfer, but only after. This seems like a bad idea now that our
average snapshot size has gone up, as placeholders help avoid duplicate work;
we now install them before accepting the range data.

In practice, we currently only allow once snapshot inflight per store, so it
is not clear that that improvement is buying us anything. I think that the
improved clarity in this commit will, though.

Concretely, we now have two simple rules for placeholders:

1. they only exist for uninitialized replicas (and must exist for those, or we
   can end up with overlapping replicas)
2. you write it --> you (and *only* you) remove it.


1 was true before but the documentation was not clear. 2. was not true, as
there were a few out-of-band places that removed placeholders (and didn't
clearly explain why - sounded like we were previously leaking placeholders,
maybe @ajwerner remembers?).

There's also a bit of extra cleanup that clarifies things - for example, since
caller to `applySnapshot` already checked for an empty snapshot, the
placeholder removal within it effectively dead code and is now removed. Also,
in receiveSnapshot, there was an optimistic check that allowed us to discard
snapshots early. It was always confusing to keep track of which one is which
and now that we've lifted the authoritative check to the top of
receiveSnapshot, we got to delete the other one.

Release note: None

tbg committed:
kvserver: simplify and document handling of ReplicaPlaceholder

This commit cleans up the handling of ReplicaPlaceholder and documents their
semantics better. I embarked on this while looking into an unrelated test
flake. It occurred to me that we were not installing placeholders before the
snapshot transfer, but only after. This seems like a bad idea now that our
average snapshot size has gone up, as placeholders help avoid duplicate work;
we now install them before accepting the range data.

In practice, we currently only allow once snapshot inflight per store, so it
is not clear that that improvement is buying us anything. I think that the
improved clarity in this commit will, though.

Concretely, we now have two simple rules for placeholders:

1. they only exist for uninitialized replicas (and must exist for those, or we
   can end up with overlapping replicas)
2. you write it --> you (and *only* you) remove it.

1. was true before but the documentation was not clear. 2. was not true, as
there were a few out-of-band places that removed placeholders (and didn't
clearly explain why - sounded like we were previously leaking placeholders).

There's also a bit of extra cleanup that clarifies things - for example,
since caller to `applySnapshot` already checked for an empty snapshot,
the placeholder removal within it effectively dead code and is now
removed. Also, in receiveSnapshot, there was an optimistic check that
allowed us to discard snapshots early. It was always confusing to keep
track of which one is which and now that we've lifted the authoritative
check to the top of receiveSnapshot, we got to delete the other one.

I ended up significantly rewriting TestReportUnreachableRemoveRace (as
it is much more flaky with the early-installed placeholders), and so
this also:

Closes #62955

as verified with 20k+ iterations of roachprod-stress on a 20 node 8cpu
cluster. As part of this I went ahead and made replication changes
carried out by TestCluster aware of retriable replication errors,
which it will now retry for up to 45s.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62719)
<!-- Reviewable:end -->

tbg commented:
Note that this is a lower-priority review as it's not relevant to the stability period.
cc @cockroachdb/kv-east in case anyone wants to learn about placeholders/snapshots and suggest improvements to the docs.

ajwerner reviewed (commented):
superficial first pass

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-:-MWyLxLb6ot1mAqZa87B:b-3maakp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_raft.go, line 545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MWyJydU8-3TCbMKEr3e:-MWyJydU8-3TCbMKEr3f:bei2ukj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_raft.go#L545)):*
> ```Go
> 	}
> 
> 	if inSnap.State != nil {
> ```

commentary on when this is nil? 
___
*[pkg/kv/kvserver/replica_raft.go, line 545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MWyKsbs4BUZ21qYbEpR:-MWyKsbs4BUZ21qYbEpS:b-6jjmsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_raft.go#L545)):*
> ```Go
> 	}
> 
> 	if inSnap.State != nil {
> ```


This method could benefit from being broken into helpers with some struct accumulating the relevant state but this doesn't seem like the time for it.
___
*[pkg/kv/kvserver/store_raft.go, line 304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MWyLW2j5acwxJg_joV2:-MWyLW2j5acwxJg_joV3:brboj7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/store_raft.go#L304)):*
> ```Go
> 			// forward, but an uninitialized replica's applied index is zero (and a
> 			// snapshot's is at least raftInitialLogIndex).
> 			if inSnap.placeholder != nil {
> ```

should we enforce the removal of a placeholder corresponds with a placeholder object by making the method for removal either take a `ReplicaPlaceholder` or maybe better be a method on a `ReplicaPlaceholder` which poisons it for future use?


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> Reviewed 10 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62719)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_placeholder.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MX1w0VJ2DGesMUAycC7:-MX1w0VJ2DGesMUAycC8:b1sm0sm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_placeholder.go#L72)):*
> This simply *results* in


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 10 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62719)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @tbg)

___
*[pkg/kv/kvserver/replica_placeholder.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MXhmP8y1n3dem_KZM5D:-MXhmP8y1n3dem_KZM5E:bg8hedm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_placeholder.go#L62)):*
> ```Go
> // - a placeholder can only be removed by the operation that installed it, and
> //   that operation *must* eventually remove it. In practice, they are inserted
> //   before receiving the snapshot data, so they are fairly long-lived.
> ```

Can you give a hint about when they are removed. Something like "... receiving the snapshot data and are replaced when applying the snapshot, so ..."
___
*[pkg/kv/kvserver/replica_placeholder.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MXhmtOn1OYm8zlRQa_d:-MXhmtOn1OYm8zlRQa_e:b-22cd3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_placeholder.go#L69)):*
> ```Go
> // can delete it).
> //
> // Note also that uninitialized replicas can get replicaGC'ed, so it is
> ```

I think it would be worth calling out somewhere how a placeholder relates to an uninitialized replica. We talk about one of the rules above, but don't complete the picture of how they relate.
___
*[pkg/kv/kvserver/replica_raft.go, line 545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MWyJydU8-3TCbMKEr3e:-MXhnPBv328VVtjpcvjm:b-ljmoqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_raft.go#L545)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

commentary on when this is nil? 
</blockquote></details>

+1. Is it possible for `inSnap.State == nil && !raft.IsEmptySnap(rd.Snapshot)`? That would be surprising, and it's not clear that ignoring `rd.Snapshot` is correct.
___
*[pkg/kv/kvserver/replica_raft.go, line 591 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MXhny82-mNxXyndI_2h:-MXhny82-mNxXyndI_2i:b-sc3crx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_raft.go#L591)):*
> ```Go
> 				refreshReason = reasonSnapshotApplied
> 			}
> 		} else if inSnap.placeholder != nil {
> ```

Was the `inSnap.State != nil` condition added to make this safe? If so, does `inSnap.placeholder != nil` not imply `inSnap.State != nil`?
___
*[pkg/kv/kvserver/replica_raft.go, line 593 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MXhnkpK1fT-jjTtJW0o:-MXhnkpK1fT-jjTtJW0p:b-v4k7g3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_raft.go#L593)):*
> ```Go
> 		} else if inSnap.placeholder != nil {
> 			// We asked raft to apply a snapshot, but it did not present it for
> 			// application.
> ```

Could you give a hint about why this might happen?
___
*[pkg/kv/kvserver/replica_raftstorage.go, line 795 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MXhoIyhE_DqR20IOvIc:-MXhoIyhE_DqR20IOvId:b-1xdvzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/527f2436edef9633fffe341c364f29b11bbe4db6/pkg/kv/kvserver/replica_raftstorage.go#L795)):*
> ```Go
> 	}()
> 
> 	if raft.IsEmptySnap(snap) {
> ```

Want to turn this into an assertion?
___
*[pkg/kv/kvserver/store_snapshot.go, line 584 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MXhp1N58B7-knUY9TLK:-MXhp1N58B7-knUY9TLL:b-9e4ctt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/store_snapshot.go#L584)):*
> ```Go
> }
> 
> // canAcceptSnapshotLocked returns (_, nil) if the snapshot can be applied. this
> ```

"to this store's"


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Updated, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-:-M_KuAWSBst9Ek_0SiRq:b-rr168b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @irfansharif)

___
*[pkg/kv/kvserver/replica_placeholder.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MXhmtOn1OYm8zlRQa_d:-M_KclzlAQ16LDQVKK01:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_placeholder.go#L69)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think it would be worth calling out somewhere how a placeholder relates to an uninitialized replica. We talk about one of the rules above, but don't complete the picture of how they relate.
</blockquote></details>

Good idea, done.
___
*[pkg/kv/kvserver/replica_placeholder.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MX1w0VJ2DGesMUAycC7:-M_KctsPBDSJnw-SjyXu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_placeholder.go#L72)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

> This simply *results* in
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_raft.go, line 545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MWyJydU8-3TCbMKEr3e:-M_Ke8OT4qtrEfQvKGHS:bqz8him) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_raft.go#L545)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1. Is it possible for `inSnap.State == nil && !raft.IsEmptySnap(rd.Snapshot)`? That would be surprising, and it's not clear that ignoring `rd.Snapshot` is correct.
</blockquote></details>

Added this as an assertion.
___
*[pkg/kv/kvserver/replica_raft.go, line 591 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MXhny82-mNxXyndI_2h:-M_KfkkwBj3yDb8QLSC2:b-5x47gj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/replica_raft.go#L591)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Was the `inSnap.State != nil` condition added to make this safe? If so, does `inSnap.placeholder != nil` not imply `inSnap.State != nil`?
</blockquote></details>

No, added a comment. We don't get a placeholder if the target replica is already initialized. Recall that inited replicas and placeholders share the replicasByKey btree.
___
*[pkg/kv/kvserver/replica_raftstorage.go, line 795 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MXhoIyhE_DqR20IOvIc:-M_Kg6lOCRONjrdcgBmB:b-mcdswi) ([raw file](https://github.com/cockroachdb/cockroach/blob/527f2436edef9633fffe341c364f29b11bbe4db6/pkg/kv/kvserver/replica_raftstorage.go#L795)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Want to turn this into an assertion?
</blockquote></details>

The parent now checks that inSnap and rd.Snapshot match up.
___
*[pkg/kv/kvserver/store_raft.go, line 304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-MWyLW2j5acwxJg_joV2:-M_KoPtx7_VSoF6BpbNM:bbm8rs3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba1c5dd1e5e1c7b2c95b3e468d8a15de70fff5e/pkg/kv/kvserver/store_raft.go#L304)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

should we enforce the removal of a placeholder corresponds with a placeholder object by making the method for removal either take a `ReplicaPlaceholder` or maybe better be a method on a `ReplicaPlaceholder` which poisons it for future use?
</blockquote></details>

Good idea, did something like that, PTAL


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-:-M_L-mJF3HxGj6_0Bldv:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62719)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/kv/kvserver/client_raft_test.go, line 2506 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-M_LsALVEt6H_IC0SrAZ:-M_LsALVEt6H_IC0SrA_:bb5dj6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4bed54c31c9a38d88f672189cc283dc848e8abd/pkg/kv/kvserver/client_raft_test.go#L2506)):*
> ```Go
> 
> outer:
> 	for i := 0; i < 50; i++ {
> ```

This failed in CI
___
*[pkg/kv/kvserver/replica_placeholder.go, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62719#-M_Lov9z89kzWHK8RBVy:-M_Lov9z89kzWHK8RBVz:bgv7zl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4bed54c31c9a38d88f672189cc283dc848e8abd/pkg/kv/kvserver/replica_placeholder.go#L67)):*
> ```Go
> //   or the snapshot was fully applied (in which case the placeholder is exchanged
> //   for a RangeDescriptor).
> // - Placeholders must not be copied (i.e. always pass by reference).
> ```

super nit: the rest of this list does not have a capitalized first letter but this one does. 


<!-- Sent from Reviewable.io -->

tbg commented:
Hmm. `TestReportUnreachableRace` is flaky now because of possible interleavings between raft snapshots and upreplication-snapshots, but behind that hides a real issue:

> F210511 10:15:31.492990 13031 kv/kvserver/store_raft.go:306  [n1,s1,r38/26:{-}] 1  unable to remove placeholder: placeholder should not exist in store: range=38 [/System/"aaa-testing"-/System/NodeLiveness) (placeholder)

Still digging into why.

tbg commented:
Still deflaking this test, but I think now I'm just looking at flakes that were always there. It seems that sometimes, when we re-add the learner and promote it to a voter, the replica doesn't catch up to that change. I suspect that there's some unexpected quiescing going on, but still need to confirm. Unfortunately, it takes well over an hour on 20+ nodes to repro at this point.

tbg committed:
kvserver: document pitfalls around DisableRaftSnapshotQueue

Disabling this might be tempting sometimes, but it makes up-replication
flaky due to corner cases in which we need a Raft snapshot to unwedge
things. I just spent a lot of time on this, so document what I learned

I had (hackily) implemented the various improvements mentioned in this
comment, but they're too dodgy for production code, and besides there
would still be more holes left that would be more difficult to plug.

Release note: None

tbg commented:
It's green!

bors r=erikgrinaker
